### ISSUE TITLE
Feature request: use only specified number of CPUs

### ISSUE THREAD

**Issue by zarniwhoop73** [rust-lang/rust#73712]
Apologies if this has been discussed before, but I can't find anything. I do a lot of compilation on desktop machines, and for some things (Beyond Linuxfromscratch) it is useful to give an idea of how long a build will take on a low-end machine - for that we use 4 cores (and our "standard build unit"). This particularly applies when building firefox and thunderbord, but I guess it is also applicable to smaller packages such as librsvg and even cbindgen..

I currently have a 4-core machine which I can use for this, but I'm hoping to replace it with something faster which will also have more cores. I know I can limit how many cores are **actually** available to the term in which I'm running a job by using the taskset command in linux. But my experience from trying that on a machine with 8 cores was that rust scheduled compiles as if all 8 were available, so the poor system was overloaded by too many jobs.

Meanwhile, when I upgrade firefox on my laptop I have 8 cores (SMT) but far too little real memory. In that case I have to take 4 cores offline to have a chance of compiling firefox without it swapping. So again, being able to tell rust that I don't want to use all the cores would be helpful.

Finally, if I do replace the current 4-core machine it occurs to me that I could run compiles as a background task while I'm doing other things, e.g. if I had 12 cores I could reserve 8 for the compiles and still have 4 available for browsers / playing videos / general housekeeping. With conventional make I could use -j8 in that situation, and similarly with ninja the number of jobs can be limited. But with rust thre does not seem any obvious way to e.g. pass an environment variable to num_cpus to say "pretend I only have this many cores".

**Comment by nagisa** [comment]
Cargo+rustc use jobserver to ensure it does not hammer the underlying system with unnecessary threads. Using that functionality, if `-j4` is specified to `cargo`, the overall build should not use more than 4 cores, at least as far as the `rustc` portion is concerned. LLVM and linkers could still use more cores, but that's mostly outside of Rust's control. Similar functionality can be achieved in similar way by calling `rustc` from different build systems.

I don’t know off-hand if its possible to supply jobserver token to cargo or not, but if it isn’t that’s a [feature request for cargo](https://github.com/rust-lang/cargo/) should be made.

How projects like Firefox integrate Rust into their build systems is entirely a concern for the project in question and would be out of scope in this repository.

**Comment by jonas-schievink** [comment]
Cargo might already inherit make's jobserver. If that doesn't work, please file a Cargo feature request.

**Comment by Mark-Simulacrum** [comment]
It is indeed the case that Cargo, and rustc, both should play well with being invoked with an existing jobserver in the environment; bugs should be filed against Cargo first if you believe that it isn't working there (you have more than N rustc processes). If you have too many threads then a new issue here would be reasonable.

**Comment by zarniwhoop73** [comment]
Thank you to all of you. I've now identified that a number of things were involved (e.g. for firefox the python mach script controls the number of jobs and now seems reliable when told to use a specified number of jobs). For compiling rustc itself, I've now found documentation on the options for x.py and I'm passing --jobs 4 on each stage. When I tried (only) that last night, it was ok for a while, and then later used all 8 cores, eventually leading to significant swap usage (and I did not run the tests).

Today I've been testing on a desktop with 8 cores and an adequate amount of memory. Using both --jobs 4 **and** exporting CARGO_BUILD_JOBS=4 mostly did the right thing. The exception was _some_ of the tests - many of them only used up to 400% CPU, but at one point some were using 800%. For me that is not catastrophic.

I'm adding this here in the hope that google might find it - searching for anything about rust seems to get matches for a game of that name and what happens to iron and steel in the wet  And inevitably most of the results which are about the rust compiler focus on questions about using **more** cores.
